2021-02-08-infinispan-kubectl.adoc

```diff
@@ -0,0 +1,56 @@
+---
+layout: blog
+title: Infinispan Kubectl
+permalink: /blog/:year/:month/:day/infinispan-12-final
+date: '2021-02-08T12:00:00.000-00:00'
+author: remerson
+tags: [ "cli", "quarkus", "native", "kubernetes"]
+---
+
+Now that the Infinispan CLI is available as a TODO[native binary] it's possible to utilise it as a kubectl client plugin.
+
+Integrating with `kubectl` lets you install and remove the Infinispan Operator, create and delete Infinispan clusters, and obtain information about various resources.
+
+== Installation
+To install the CLI as a `kubectl` plugin, do the following:
+
+. Copy the `ispn-cli`, or create a symbolic link, to a file named "kubectl-infinispan".
+
+. Add `kubectl-infinispan` to your `PATH`.
+
+. Run `kubectl plugin list` to verify the CLI is added as a plugin.
+
+. Run `kubectl infinispan --help` to start using the CLI plugin.
+
+For more detailed instructions, please see the https://kubernetes.io/docs/tasks/extend-kubectl/kubectl-plugins/#installing-kubectl-plugins[kubectl docs].
+
+== Examples
+Before executing the examples login to your Kubernetes cluster with `kubectl login`.
+
+==== Installing Infinispan Operator
+----
+kubectl infinispan install
+----
+
+==== Creating  Infinispan Clusters
+----
+kubectl infinispan create --replicas=2 myinfinispan
+kubectl infinispan ls
+----
+
+==== Removing Clusters
+----
+kubectl infinispan rm myinfinispan
```

Comment:
shouldn't be `delete` instead of `rm`?
How can I test this? I guess using the `ispn-cli` binary isn't the right way:
```
$ kubectl plugin list 
The following compatible plugins are available:

/tmp/kubeispn/kubectl-infinispan
$ kubectl infinispan install
cli 'install' is not a cli command. See 'cli --help'.
Usage: cli [<options>]
...
```

Reply:
For some reason it seems like the `cli` isn't detecting that it's been executed within `kubectl`.

What happens when you execute `kubectl infinispan -h`?

I get the following:

```
kube 'help' is not a kube command. See 'kube --help'.
Usage: kube
Kubernetes commands

kube commands:
    create     Creates a resource
    install    Installs the operator
    shell      Initiates an interactive shell with a service
    get        Displays resources
    delete     Deletes a resource
    uninstall  Uninstalls the operator
    version    Shows version information
```

Reply:
I'm doing the following steps:
```
$ unzip infinispan-cli-12.0.0.Final-linux-amd64.zip 
Archive:  infinispan-cli-12.0.0.Final-linux-amd64.zip
   creating: infinispan-quarkus-cli-12.0.0.Final/
  inflating: infinispan-quarkus-cli-12.0.0.Final/LICENSE.md  
  inflating: infinispan-quarkus-cli-12.0.0.Final/ispn-cli  
  inflating: infinispan-quarkus-cli-12.0.0.Final/README.md  

$ ln -s infinispan-quarkus-cli-12.0.0.Final/ispn-cli kubectl-infinispan

$ export PATH=$PWD:$PATH

$ kubectl plugin list
The following compatible plugins are available:

/tmp/kubetest/kubectl-infinispan

$ kubectl infinispan -h
Usage: cli [<options>]


Options:
  -t, --truststore           A truststore to use when connecting to SSL/TLS-enabled servers
  -s, --truststore-password  The password for the truststore
  -v, --version              Shows version information
  --trustall                 Whether to trust all certificates
  -f, --file                 File for batch mode
  -c, --connect              A connection URL. Use '-' to connect to http://localhost:11222
  -P, --properties           Sets system properties from the specified file.
  -D, --propertyMap          Sets a system property
  -h, --help

cli commands:
    add          Adds/subtracts a value to/from a counter
    backup       Manages container backup creation and restoration
    benchmark    Benchmarks server performance
    cache        Selects a cache
    cas          Compares and sets counter values
    cd           Selects a subsystem or item
    clear        Clears the screen
    clearcache   Clears the cache
    config       Configuration operations
    connect      Connects to a remote server
    container    Selects a container
    counter      Selects counters
    create       Creates a cache or a counter
    credentials  Credential store operations
    describe     Displays information about the specified resource
    disconnect   Disconnects from a remote server
    drop         Drops a cache or a counter
    echo         Echoes messages to the output. Useful for adding information to batch runs.
    encoding     Gets/sets the current encoding
    get          Gets an entry from the cache
    logging      Inspects/Manipulates the server logging configuration
    ls           Lists resources in a path
    migrate      Migration operations
    patch        Patch operations
    put          Puts an entry into the cache
    query        Queries a cache
    remove       Removes an entry from the cache
    reset        Resets a counter to its initial value
    run          Reads and executes commands from one or more files
    schema       Manipulates protobuf schemas
    server       Obtains information about the server
    shutdown     Shuts down individual servers or the entire cluster
    site         Manages backup sites
    task         Executes or manipulates server-side tasks
    user         User operations
    version      Shows version information
```

I don't have minikube cluster running, if that matters.

Reply:
```
ln -s infinispan-quarkus-cli-12.0.0.Final/ispn-cli kubectl-infinispan
```

Symbolic links don't work unfortunately. I should probably add that to the blog.

Reply:
I just realised that I explicitly say to create a symbolic link :sweat_smile: ... fixing.